### PR TITLE
Since default is dynbinary we should test with dynbinary

### PR DESCRIPTION
**Pull request by rhatdan**
Signed-off-by: Dan Walsh dwalsh@redhat.com

### CODE REVIEW

**Comment by cpuguy83** [comment]
ping @tianon @jfrazelle

**Comment by LK4D4** [comment]
tgz is failing

File: hack/make.sh [hunk 1]
```diff
@@ -52,14 +52,12 @@ DEFAULT_BUNDLES=(
 	validate-toml
 	validate-vet
 
-	binary
```
Comment (jfrazelle) [inline hack/make.sh]: we should keep both here

Comment (rhatdan) [inline hack/make.sh]: Yes we probably should

File: Makefile [hunk 2]
```diff
@@ -50,25 +50,28 @@ binary: build
...
 cross: build
-	$(DOCKER_RUN_DOCKER) hack/make.sh binary cross
+	$(DOCKER_RUN_DOCKER) hack/make.sh dynbinary binary cross
 
 deb: build
-	$(DOCKER_RUN_DOCKER) hack/make.sh binary build-deb
+	$(DOCKER_RUN_DOCKER) hack/make.sh dynbinary build-deb
+
+dynbinary: build
+	$(DOCKER_RUN_DOCKER) hack/make.sh dynbinary
```
Comment (hqhq) [inline Makefile]: I think dynbinary is not encouraged on this level, see https://github.com/docker/docker/pull/12224 .

Comment (rhatdan) [inline Makefile]: We already run distro specific tests and builds.  But doesn't this docker run docker in docker?

Comment (runcom) [inline Makefile]: I've seen this introduced here as well https://github.com/diogomonica/docker/commit/50852cdc4d7ecfdab795018309aca783880ec12d#diff-b67911656ef5d18c4ae36cb6741b7965R51
PR is https://github.com/docker/docker/pull/17937

Comment (jfrazelle) [inline Makefile]: yes its a nice to have

Comment (tianon) [inline Makefile]: If you want to build a binary for your host from within the Docker
container, the only sure solution is the static binary.

Comment (jfrazelle) [inline Makefile]: Yeah just for testing its nice






On Thu, Nov 12, 2015 at 7:47 AM -0800, "Tianon Gravi" <notifications@github.com> wrote:












In Makefile:
>  
>  deb: build
> -	$(DOCKER_RUN_DOCKER) hack/make.sh binary build-deb
> +	$(DOCKER_RUN_DOCKER) hack/make.sh dynbinary build-deb
> +
> +dynbinary: build
> +	$(DOCKER_RUN_DOCKER) hack/make.sh dynbinary

If you want to build a binary for your host from within the Docker
container, the only sure solution is the static binary.

Comment (jfrazelle) [inline Makefile]: But we don't _need_ if people get confused that's the last thing we need lol






On Thu, Nov 12, 2015 at 7:47 AM -0800, "Tianon Gravi" <notifications@github.com> wrote:












In Makefile:
>  
>  deb: build
> -	$(DOCKER_RUN_DOCKER) hack/make.sh binary build-deb
> +	$(DOCKER_RUN_DOCKER) hack/make.sh dynbinary build-deb
> +
> +dynbinary: build
> +	$(DOCKER_RUN_DOCKER) hack/make.sh dynbinary

If you want to build a binary for your host from within the Docker
container, the only sure solution is the static binary.

**Comment by LK4D4** [comment]
I don't fully understand `dynbinary` bikeshed now, but LGTM

File: Makefile (hunk 2 above)
Comment (tianon) [inline Makefile]: Yeah, I think people will get confused with this one and try running it to get a binary for their host system, so I'd rather we leave this one as-is.

Comment (tianon) [inline Makefile]: (meaning `make binary` which creates a static binary)

**Comment by LK4D4** [comment]
@tianon You're proposing to close or not to close? :)

**Comment by tianon** [comment]
I'm proposing a minor change -- other than that one Make target, this looks
fine. :+1:

**Comment by rhatdan** [comment]
What would you like to see?

make 

Default to dynbinary and add make static build a static version?

**Comment by tianon** [comment]
I would prefer to see `make dynbinary` not exist since the `Makefile` is a convenience for users.  If they want to test `dynbinary` by itself, that should be done indirectly through one of the test targets like you've got them here (which is perfect, :+1:), or directly via `make shell` followed by `./hack/make.sh dynbinary` -- so `make` would default to `make binary` same as it does today.

**Comment by tianon** [comment]
Otherwise, we'll get lots of reports from users on non-Ubuntu systems who ran `make dynbinary` and couldn't successfully run the resulting binary on their host system.

**Comment by rhatdan** [comment]
Well we are building on non ubuntu machines and make dynbinary is working for us.  Is this the case of them copying the actual binary somewhere else?

**Comment by tianon** [comment]
`make dynbinary` itself does "work", in that it compiles a binary -- that binary will be in `bundles/VERSION` on the host system, which is all still normal.  The problem comes in when a user does `make dynbinary` and then tries to run the binary inside `bundles` on their host, since it's going to be linked against the specific ABIs used in the Ubuntu image it was built in, not whatever their host system is.

**Comment by rhatdan** [comment]
Well I guess this goes back to other patches which we have proposed to use a Dockerfile based on the OS you are testing against.

We use Dockerfiles for RHEL on RHEL, Fedora on Fedora and Centos on Centos.

**Comment by tianon** [comment]
Right, but then you have to have a separate "Dockerfile" for each _version_
of each distro as well.  This is how we build our ".deb" and ".rpm" files
now; see https://github.com/docker/docker/tree/master/contrib/builder for
the scripts that generate the bazillion Dockerfiles involved.

The way I personally use "dynbinary" is via "AUTO_GOPATH=1 ./hack/make.sh
dynbinary" directly on my host without using the Makefile or Dockerfile at
all, the same way our packagers do.  As long as your host system has the
proper build dependencies (which the libraries of are required for running
anyhow), this works really well.

**Comment by rhatdan** [comment]
Same with us.

**Comment by rhatdan** [comment]
I am not interested in using the Makefile for building dybinary for personal use, only to make sure make test and friends use dynbinary for testing.

**Comment by tianon** [comment]
Right, I agree completely; that's exactly why I don't want an explicit `dynbinary` target, but _do_ like having the other `test` targets use `dynbinary` (since they're reasonably safe in doing so).

**Comment by LK4D4** [comment]
@tianon Could you illustrate your proposal with patch? I agree that dynbinary is sorta useless target.

**Comment by tianon** [comment]
Heh, sure:
```patch
diff --git a/Makefile b/Makefile
index 47f3c1e..fa8c1c4 100644
--- a/Makefile
+++ b/Makefile
@@ -1,4 +1,4 @@
-.PHONY: all binary build cross default docs docs-build docs-shell dynbinary shell test test-unit test-integration-cli test-docker-py validate
+.PHONY: all binary build cross default docs docs-build docs-shell shell test test-unit test-integration-cli test-docker-py validate
 
 # env vars passed through directly to Docker's build scripts
 # to allow things like `make DOCKER_CLIENTONLY=1 binary` easily
@@ -55,9 +55,6 @@ cross: build
 deb: build
 	$(DOCKER_RUN_DOCKER) hack/make.sh dynbinary build-deb
 
-dynbinary: build
-	$(DOCKER_RUN_DOCKER) hack/make.sh dynbinary
-
 rpm: build
 	$(DOCKER_RUN_DOCKER) hack/make.sh dynbinary build-rpm
 
```

**Comment by LK4D4** [comment]
@rhatdan Would you mind to do this and we'll merge your PR :)

**Comment by rhatdan** [comment]
Ok done.

@nalind PTAL

**Comment by tianon** [comment]
LGTM

**Comment by LK4D4** [comment]
LGTM

**Pull request merged**